File: src/main/java/ukalus/path/astar/AStar.java
package ukalus.path.astar;

import java.util.*;

/**
 * @author gremopm
 */
public class AStar {

  public Iterator<Node> solve(Heuristic heuristic, Cost cost, Node start, Node stop) {
    if (start == null || stop == null) {
      throw new IllegalArgumentException();
    }

    Queue<Node> open = new PriorityQueue<>(1, new NodeTotalComparator());
    List<Node> close = new LinkedList<>();
    Node current = start;
    open.add(start);
    while (!open.isEmpty() && !current.equals(stop)) {
      current = open.poll();
      for (Node successor : current.getSuccessors()) {
        int newCost = cost.calculate(current, successor);
        Node openNode = open.stream().filter(successor::equals).findFirst().orElse(null);
        if (openNode == null || openNode.getCost() > newCost) {
          Node closeNode = close.stream().filter(successor::equals).findFirst().orElse(null);
          if (closeNode == null || closeNode.getCost() > newCost) {
            successor.setCost(newCost);
            successor.setEstimate(heuristic.estimate(successor));
            open.remove(openNode);
            close.remove(closeNode);
            open.add(successor);
          }
        }
      }
      close.add(current);
    }

    LinkedList<Node> result = new LinkedList<>();
    while (current != null) {
      result.addFirst(current);
      current = current.getParent();
    }

    return result.iterator();
  }
}

File: src/main/kotlin/ukalus/path/astar/AStar.kt
package ukalus.path.astar

import java.util.*
import java.util.Comparator.comparingDouble

/**
 * @author gremopm
 */
class AStar {
    fun solve(heuristic: (Node) -> Double, cost: (Node, Node) -> Double, start: Node, stop: Node): List<Node> {
        val open = PriorityQueue<Node>(1, comparingDouble(Node::cost)).apply { add(start) }
        val close = mutableListOf<Node>()
        var current = start
        while (!open.isEmpty() && current != stop) {
            current = open.poll()!!
            for (successor in current.successors) {
                val newCost = cost(current, successor)
                if (successor !in open && successor !in close || successor.cost > newCost) {
                    successor.cost = newCost
                    successor.estimate = heuristic(successor)
                    close.remove(successor)
                    open.add(successor)
                }
            }
            close.add(current)
        }
        return if (current == stop)
            generateSequence(current) { it.parent }.fold(mutableListOf()) { s, x -> s.apply { add(0, x) } }
        else
            emptyList()
    }
}